LWIP without OS-awareness produces a compile errors

This software library is a masterpiece, 😎 and we decided to use it under extreme conditions on a minimalist IoT controller. However, we encountered an unexpected problem: it's impossible to compile libcoap for the tiny STM32 with LWIP 2.1.2 but without OS-awareness (i.e., without FreeRTOS). 😔

The functions and constants `getaddrinfo()`, `AF_UNSPEC`, `AF_INET` etc. are undefined when a LWIP is used (`WITH_LWIP` = 1) but without an OS-awareness (`NO_SYS` = 1) and dependly without a Sockets API (`LWIP_SOCKET` = 0). The first compile error goes there: `coap_address.c:coap_resolve_address_info()`. Moreover the `struct addrinfo` can be also undefined by `LWIP_DNS` = 0.

As a quick and dirty workaround, the following code can be added to the user's  `coap_config.h` file:

```
#if WITH_LWIP && NO_SYS

#define AF_UNSPEC       0
#define AF_INET         2

#define getaddrinfo(nodname, servname, hints, res) \
   lwip_getaddrinfo(nodname, servname, hints, res)

typedef uint32_t in_addr_t;

struct in_addr {
  in_addr_t s_addr;
};

struct addrinfo {
  int               ai_flags;      /* Input flags. */
  int               ai_family;     /* Address family of socket. */
  int               ai_socktype;   /* Socket type. */
  int               ai_protocol;   /* Protocol of socket. */
  uint32_t          ai_addrlen;    /* Length of socket address. */
  struct sockaddr  *ai_addr;       /* Socket address of socket. */
  char             *ai_canonname;  /* Canonical name of service location. */
  struct addrinfo  *ai_next;       /* Pointer to next in list. */
};

struct sockaddr_in {
  uint8_t         sin_len;
  uint8_t         sin_family;
  uint16_t        sin_port;
  struct in_addr  sin_addr;
#define SIN_ZERO_LEN 8
  char            sin_zero[SIN_ZERO_LEN];
};

#endif /* WITH_LWIP && NO_SYS */
```
By the way, the `check_function_exists(getaddrinfo HAVE_GETADDRINFO)` call is used in the CMakeLists.txt file, but the result is not used anywhere.

I humbly hope that someone will be able to adapt the library for the above conditions in the right way.

Although all related issues have been resolved, the main incompatibility issue still requires attention.

I will try to find a way to implement a limited version of the `getaddrinfo()` function (i.e. without DNS) to avoid disabling this functionality completely, as has been done in other parts of libcoap for LWIP.

Example of an error message if socket support is disabled:
```c
**** Incremental Build of configuration Debug for project 429_CoAP ****

make -j4 all 

arm-none-eabi-gcc "../libcoap-develop/src/coap_address.c" -mcpu=cortex-m4 -std=gnu11 -g3 -DDEBUG -DLWIP_DEBUG -DUSE_HAL_DRIVER -DSTM32F429xx -c -I../Core/Inc -I../Drivers/STM32F4xx_HAL_Driver/Inc -I../Drivers/STM32F4xx_HAL_Driver/Inc/Legacy -I../Drivers/CMSIS/Device/ST/STM32F4xx/Include -I../Drivers/CMSIS/Include -I../LWIP/App -I../LWIP/Target -I../Middlewares/Third_Party/LwIP/src/include -I../Middlewares/Third_Party/LwIP/system -I../Drivers/BSP/Components/lan8742 -I../Middlewares/Third_Party/LwIP/src/include/netif/ppp -I../Middlewares/Third_Party/LwIP/src/include/lwip -I../Middlewares/Third_Party/LwIP/src/include/lwip/apps -I../Middlewares/Third_Party/LwIP/src/include/lwip/priv -I../Middlewares/Third_Party/LwIP/src/include/lwip/prot -I../Middlewares/Third_Party/LwIP/src/include/netif -I../Middlewares/Third_Party/LwIP/src/include/compat/posix -I../Middlewares/Third_Party/LwIP/src/include/compat/posix/arpa -I../Middlewares/Third_Party/LwIP/src/include/compat/posix/net -I../Middlewares/Third_Party/LwIP/src/include/compat/posix/sys -I../Middlewares/Third_Party/LwIP/src/include/compat/stdc -I../Middlewares/Third_Party/LwIP/system/arch -I../libcoap-develop/include -O0 -ffunction-sections -fdata-sections -Wall -fstack-usage -MMD -MP -MF"libcoap-develop/src/coap_address.d" -MT"libcoap-develop/src/coap_address.o" --specs=nano.specs -mfpu=fpv4-sp-d16 -mfloat-abi=hard -mthumb -o "libcoap-develop/src/coap_address.o"

../libcoap-develop/src/coap_address.c: In function 'coap_resolve_address_info':
../libcoap-develop/src/coap_address.c:588:19: error: storage size of 'hints' isn't known
  588 |   struct addrinfo hints;
      |                   ^~~~~
../libcoap-develop/src/coap_address.c:643:11: warning: implicit declaration of function 'getaddrinfo' [-Wimplicit-function-declaration]
  643 |   error = getaddrinfo(addrstr, NULL, &hints, &res);
      |           ^~~~~~~~~~~
In file included from ../libcoap-develop/include/coap3/coap_net.h:44,
                 from ../libcoap-develop/include/coap3/coap_async.h:20,
                 from ../libcoap-develop/include/coap3/coap.h:43,
                 from ../libcoap-develop/include/coap3/coap_internal.h:83,
                 from ../libcoap-develop/include/coap3/coap_libcoap_build.h:45,
                 from ../libcoap-develop/src/coap_address.c:16:
../libcoap-develop/src/coap_address.c:646:53: warning: implicit declaration of function 'gai_strerror'; did you mean 'strerror'? [-Wimplicit-function-declaration]
  646 |     coap_log_warn("getaddrinfo: %s: %s\n", addrstr, gai_strerror(error));
      |                                                     ^~~~~~~~~~~~
../libcoap-develop/include/coap3/coap_debug.h:292:30: note: in definition of macro 'coap_log'
  292 |       coap_log_impl((level), __VA_ARGS__); \
      |                              ^~~~~~~~~~~
../libcoap-develop/src/coap_address.c:646:5: note: in expansion of macro 'coap_log_warn'
  646 |     coap_log_warn("getaddrinfo: %s: %s\n", addrstr, gai_strerror(error));
      |     ^~~~~~~~~~~~~
../libcoap-develop/src/coap_address.c:646:19: warning: format '%s' expects argument of type 'char *', but argument 4 has type 'int' [-Wformat=]
  646 |     coap_log_warn("getaddrinfo: %s: %s\n", addrstr, gai_strerror(error));
      |                   ^~~~~~~~~~~~~~~~~~~~~~~           ~~~~~~~~~~~~~~~~~~~
      |                                                     |
      |                                                     int
../libcoap-develop/include/coap3/coap_debug.h:292:30: note: in definition of macro 'coap_log'
  292 |       coap_log_impl((level), __VA_ARGS__); \
      |                              ^~~~~~~~~~~
../libcoap-develop/src/coap_address.c:646:5: note: in expansion of macro 'coap_log_warn'
  646 |     coap_log_warn("getaddrinfo: %s: %s\n", addrstr, gai_strerror(error));
      |     ^~~~~~~~~~~~~
../libcoap-develop/src/coap_address.c:646:38: note: format string is defined here
  646 |     coap_log_warn("getaddrinfo: %s: %s\n", addrstr, gai_strerror(error));
      |                                     ~^
      |                                      |
      |                                      char *
      |                                     %d
../libcoap-develop/src/coap_address.c:650:49: error: invalid use of undefined type 'struct addrinfo'
  650 |   for (ainfo = res; ainfo != NULL; ainfo = ainfo->ai_next) {
      |                                                 ^~
../libcoap-develop/src/coap_address.c:656:18: error: invalid use of undefined type 'struct addrinfo'
  656 |     switch (ainfo->ai_family) {
      |                  ^~
../libcoap-develop/src/coap_address.c:675:24: error: invalid use of undefined type 'struct addrinfo'
  675 |           switch (ainfo->ai_family) {
      |                        ^~
../libcoap-develop/src/coap_address.c:682:48: error: invalid use of undefined type 'struct addrinfo'
  682 |             sock4 = (struct sockaddr_in *)ainfo->ai_addr;
      |                                                ^~
In file included from ../Core/Inc/coap_config.h:17,
                 from ../libcoap-develop/include/coap3/coap_libcoap_build.h:30,
                 from ../libcoap-develop/src/coap_address.c:16:
../libcoap-develop/src/coap_address.c:683:42: error: invalid use of undefined type 'struct sockaddr_in'
  683 |             info->addr.port = ntohs(sock4->sin_port);
      |                                          ^~
../Middlewares/Third_Party/LwIP/src/include/lwip/def.h:98:34: note: in definition of macro 'lwip_ntohs'
   98 | #define lwip_ntohs(x) lwip_htons(x)
      |                                  ^
../libcoap-develop/src/coap_address.c:683:31: note: in expansion of macro 'ntohs'
  683 |             info->addr.port = ntohs(sock4->sin_port);
      |                               ^~~~~
../libcoap-develop/src/coap_address.c:684:44: error: invalid use of undefined type 'struct sockaddr_in'
  684 |             memcpy(&info->addr.addr, &sock4->sin_addr, 4);
      |                                            ^~
../libcoap-develop/src/coap_address.c:739:3: warning: implicit declaration of function 'freeaddrinfo' [-Wimplicit-function-declaration]
  739 |   freeaddrinfo(res);
      |   ^~~~~~~~~~~~
../libcoap-develop/src/coap_address.c:588:19: warning: unused variable 'hints' [-Wunused-variable]
  588 |   struct addrinfo hints;
      |                   ^~~~~
make: *** [libcoap-develop/src/subdir.mk:130: libcoap-develop/src/coap_address.o] Error 1
make: *** Waiting for unfinished jobs....
"make -j4 all" terminated with exit code 2. Build might be incomplete.

Build Failed. 9 errors, 5 warnings. (took 2s.313ms)
```

--- a/src/coap_address.c
+++ b/src/coap_address.c
@@ -573,6 +573,8 @@ update_coap_addr_port(coap_uri_scheme_t scheme, coap_addr_info_t *info,
   }
 }
 
+#if defined(WITH_LWIP) && !(LWIP_DNS && LWIP_SOCKET && LWIP_COMPAT_SOCKETS)
+
 coap_addr_info_t *
 coap_resolve_address_info(const coap_str_const_t *address,
                           uint16_t port,
@@ -582,7 +584,58 @@ coap_resolve_address_info(const coap_str_const_t *address,
                           int ai_hints_flags,
                           int scheme_hint_bits,
                           coap_resolve_type_t type) {
-#if !defined(RIOT_VERSION) && !defined(WITH_CONTIKI)
+  ip_addr_t addr_ip;
+  coap_addr_info_t *info = NULL;
+  coap_addr_info_t *info_prev = NULL;
+  coap_addr_info_t *info_list = NULL;
+  coap_uri_scheme_t scheme;
+
+  (void)ai_hints_flags;
+
+  if (address == NULL || address->length == 0) {
+    memset(&addr_ip, 0, sizeof(addr_ip));
+  } else {
+    if (ipaddr_aton((const char *)address->s, &addr_ip) <= 0) {
+      coap_log_err("coap_resolve_address_info: Unable to parse '%s'\n", address->s);
+      return NULL;
+    }
+  }
+  for (scheme = 0; scheme < COAP_URI_SCHEME_LAST; scheme++) {
+    if (scheme_hint_bits & (1 << scheme)) {
+      info = get_coap_addr_info(scheme);
+      if (info == NULL) {
+        continue;
+      }
+
+      /* Need to return in same order as getaddrinfo() */
+      if (!info_prev) {
+        info_list = info;
+        info_prev = info;
+      } else {
+        info_prev->next = info;
+        info_prev = info;
+      }
+
+      memcpy(&info->addr.addr, &addr_ip, sizeof(info->addr.addr));
+
+      update_coap_addr_port(scheme, info, port, secure_port, ws_port,
+                            ws_secure_port, type);
+    }
+  }
+  return info_list;
+}
+
+#elif !defined(RIOT_VERSION) && !defined(WITH_CONTIKI)
+
+coap_addr_info_t *
+coap_resolve_address_info(const coap_str_const_t *address,
+                          uint16_t port,
+                          uint16_t secure_port,
+                          uint16_t ws_port,
+                          uint16_t ws_secure_port,
+                          int ai_hints_flags,
+                          int scheme_hint_bits,
+                          coap_resolve_type_t type) {
 
   struct addrinfo *res, *ainfo;
   struct addrinfo hints;
@@ -643,7 +696,11 @@ coap_resolve_address_info(const coap_str_const_t *address,
   error = getaddrinfo(addrstr, NULL, &hints, &res);
 
   if (error != 0) {
+#if defined(WITH_LWIP)
+    coap_log_warn("getaddrinfo: %s: %d\n", addrstr, error);
+#else /* ! WITH_LWIP */
     coap_log_warn("getaddrinfo: %s: %s\n", addrstr, gai_strerror(error));
+#endif /* ! WITH_LWIP */
     return NULL;
   }
 
@@ -738,10 +795,21 @@ coap_resolve_address_info(const coap_str_const_t *address,
 
   freeaddrinfo(res);
   return info_list;
+}
 
 #elif defined(RIOT_VERSION)
 
 #include "net/utils.h"
+
+coap_addr_info_t *
+coap_resolve_address_info(const coap_str_const_t *address,
+                          uint16_t port,
+                          uint16_t secure_port,
+                          uint16_t ws_port,
+                          uint16_t ws_secure_port,
+                          int ai_hints_flags,
+                          int scheme_hint_bits,
+                          coap_resolve_type_t type) {
 #if COAP_IPV6_SUPPORT
   ipv6_addr_t addr_ipv6;
 #endif /* COAP_IPV6_SUPPORT */
@@ -821,10 +889,21 @@ coap_resolve_address_info(const coap_str_const_t *address,
     }
   }
   return info_list;
+}
 
 #elif defined(WITH_CONTIKI)
 
 #include <os/net/ipv6/uiplib.h>
+
+coap_addr_info_t *
+coap_resolve_address_info(const coap_str_const_t *address,
+                          uint16_t port,
+                          uint16_t secure_port,
+                          uint16_t ws_port,
+                          uint16_t ws_secure_port,
+                          int ai_hints_flags,
+                          int scheme_hint_bits,
+                          coap_resolve_type_t type) {
   uip_ipaddr_t addr_ip;
   coap_addr_info_t *info = NULL;
   coap_addr_info_t *info_prev = NULL;
@@ -843,7 +922,7 @@ coap_resolve_address_info(const coap_str_const_t *address,
     }
 #endif /* COAP_IPV6_SUPPORT */
 #if COAP_IPV4_SUPPORT
-    if (family == AF_UNSPEC &&
+    if (!parsed_ip &&
         uiplib_ip4addrconv((const char *)address->s, (uip_ip4addr_t *)&addr_ip) > 0) {
       parsed_ip = 1;
     }
@@ -876,11 +955,25 @@ coap_resolve_address_info(const coap_str_const_t *address,
     }
   }
   return info_list;
-#else
-#bad OS type not supported
-  return NULL;
-#endif
 }
+
+#else
+
+#error "OS type not supported"
+
+coap_addr_info_t *
+coap_resolve_address_info(const coap_str_const_t *address,
+                          uint16_t port,
+                          uint16_t secure_port,
+                          uint16_t ws_port,
+                          uint16_t ws_secure_port,
+                          int ai_hints_flags,
+                          int scheme_hint_bits,
+                          coap_resolve_type_t type) {
+  return NULL;
+}
+
+#endif
 
 void
 coap_free_address_info(coap_addr_info_t *info) {

